### ISSUE TITLE
Add SFT to Torch Titan

### ISSUE THREAD

**Issue by abhash-er** [OpenEuroLLM/Taskboard#167]
By adding SFT support, we can leverage TorchTitan’s 3D parallelism (FSDP2 /TP / CP) for instruction tuning, alignment task and reasoning tasks. Key implementation features on top of a general SFT loop would be-

a) Packed Sequence Loading with Masking: Implement a collator that packs variable-length instruction-response pairs into fixed blocks. Crucially, this must support label_mask to ensure loss is calculated only on assistant responses, not user prompts (as mentioned in this [Issue](https://github.com/OpenEuroLLM/Taskboard/issues/166)).

b) **Smart Batching & Auto-Accumulation**: Add a `BatchSizeConfig` utility that automatically calculates the optimal `micro_batch_size` and `gradient_accumulation_steps` based on detected hardware limits (e.g., H100 vs B200) to hit the target global batch size without OOM errors.

c) **Adaptive Context Parallelism**: Logic to dynamically switch CP strategies (e.g., from _ZigZag_ to _Ring/Llama3_ style) when packed datasets are detected, ensuring that attention masking correctly respects document boundaries across distributed ranks.

d) **Hybrid Checkpoint Recovery**: A robust loading mechanism that automatically detects context: it should strictly load weights-only when initializing from a pre-trained base model, but fully load optimizer states if resuming a crashed SFT run.

**Comment by abhash-er** [comment]
>See Issue #166 for related work
>Check the [post-training ](https://github.com/OpenEuroLLM/titan-oellm/tree/post-training)branch of titan-oellm.

Check a small report (summary from claude) for the repository and some important things if anyone wants to create a new model with torch titan!

## 1. Summary of Changes 
 **`post-training` turns the repo into an SFT-capable
trainer** — adding an SFT data pipeline (with packing), the OLMo3 model family, and
ring-attention context parallelism. 

### 1.1 SFT (supervised fine-tuning) — new capability

| File | What it adds |
|---|---|
| `titan_oellm/scripts/sft.py` | SFT training entrypoint / trainer |
| `titan_oellm/datasets/sft_dataloader.py` | SFT data stack: instruction formatting + packing |
| `titan_oellm/configs/oellm_job_config.py` | Config fields: `packing_strategy`, `prepacked_dataset`, `sequential_index_cache_path`, instruction-format options |
| `titan_oellm/configs/sft_config.toml` | Example SFT config |
| `titan_oellm/scripts/run_reasoning_smoke.py` | SFT / reasoning smoke test |

**Instruction formats** (auto-detected): alpaca, chatml, sharegpt, prompt_completion,
text, plus reasoning formats (cot, reasoning_steps, problem_solution).

**Three packing strategies** (see `sft_dataloader.py`):

| Strategy | Config | Class | Notes |
|---|---|---|---|
| None (pad) | `packing_strategy = "none"` | — | Pads each sample to `seq_len` |
| Online OBFD | `packing_strategy = "obfd"` | `GlobalOBFDPackedDataset` | Global best-fit-decreasing built in-process each run (linear-scan BFD) |
| Sequential | `packing_strategy = "sequential"` | `SequentialPackedSFTDataset` | OLMo3-style greedy in-order packing; index cacheable |
| Prepacked | `prepacked_dataset = true` | `PrepackedInstructionDataset` | Reads offline-packed rows from disk; online packing force-disabled |

**Offline prepacking** (`no_push_scripts/dataset_packer.py`, *currently untracked*):
tokenizes + OBFD-packs once using a segment-tree BFD (O(log seq_len) per fit), writes
fixed-`seq_len` rows to disk in `chunked` (`.bin`/`.idx`) or `npy` format, plus
`packing_metadata.json`. Training then just mmaps the result — zero packing at train time.

> ⚠️ Prepacked output is tied to the `seq_len` and tokenizer used at pack time; they must
> match the training config. `prepacked_dataset = true` requires `packing_strategy = "none"`.

### 1.2 OLMo3 model — new architecture

`titan_oellm/models/olmo3_custom/`:
- `model/model.py` — full model (YaRN RoPE, sliding-window attention + MFU accounting, Liger fused kernels)
- `model/args.py` — `Olmo3CustomModelArgs`
- `infra/parallelize.py` — `parallelize_olmo3_custom` (incl. ring-varlen CP)
- `model/state_dict_adapter.py` — HF ⇄ titan checkpoint conversion
- `__init__.py` — flavor configs (`debugmodel`, `7B`, `32B`) + `register_train_spec`

### 1.3 Ring-attention / context parallelism

- `titan_oellm/components/flash_attention.py` and the ring-attn paths wire ring-varlen CP
  and Liger fused cross-entropy.
- Zigzag varlen forward+backward made `torch.compile`-safe.

### 1.4 Supporting fixes

- Renamed `sci_job_config` → `oellm_job_config` references.
- Prevented `None` config values from clobbering OLMo flavor defaults.
- Registered the `ring_flash_attn` custom op **before** `parallelize` imports it.
- Removed `use_flash_attn`; varlen collate now derived from `attn_type`.
- Minor `gpt_plus` (sharding plan, PP safechecks), `sci_dataloader`, tokenizer touches.

### 1.5 Qwen3 SFT + CP enablement (commit `9f38c7e`)

Qwen3 was run through the same post-training checklist OLMo3 established, so it now works
with the SFT + context-parallel pipeline. Touches `qwen3_custom/infra/parallelize.py`,
`model/args.py`, `model/model.py`, and `scripts/sft.py`. See the case study in §2.5.

---

## 2. Adding a new model

Models live under `titan_oellm/models/<name>_custom/` and self-register via a
`TrainSpec`. Use `olmo3_custom/` as the reference implementation.

### 2.1 Directory layout to create

```
titan_oellm/models/<name>_custom/
├── __init__.py                 # flavor configs + register_train_spec  ← the entry point
├── model/
│   ├── __init__.py
│   ├── args.py                 # <Name>ModelArgs dataclass
│   ├── model.py                # nn.Module; defines forward + init_weights
│   └── state_dict_adapter.py   # HF ⇄ titan checkpoint mapping
└── infra/
    ├── __init__.py
    └── parallelize.py          # parallelize_<name>_custom(model, parallel_dims, ...)
```

### 2.2 Wire-up steps

1. **Implement the model** (`model/model.py`, `model/args.py`).
   Match the existing module surface: a model class with `forward`, `init_weights`,
   and args as a dataclass subclassing the torchtitan `BaseModelArgs` convention.

2. **Implement parallelization** (`infra/parallelize.py`): a
   `parallelize_<name>_custom(model, parallel_dims, job_config, ...)` returning the
   sharded model. Cover the parallelism axes you intend to support (TP / FSDP / CP / PP).

3. **Implement the state-dict adapter** (`model/state_dict_adapter.py`) so HF checkpoints
   can be loaded/saved.

4. **Register the train spec** in `<name>_custom/__init__.py`:
   - Define a `<name>_custom_configs` dict of flavors (`debugmodel`, `7B`, …) using your args.
   - Implement `get_train_spec() -> TrainSpec` wiring `model_cls`, `model_args`,
     `parallelize_fn`, `pipelining_fn`, optimizer / lr / dataloader / tokenizer / loss /
     validator / metrics builders, and `state_dict_adapter`.
   - Call `register_train_spec("<name>_custom", get_train_spec())`.

5. **Add the import** to `titan_oellm/models/__init__.py` so the spec registers on startup:
   ```python
   from . import <name>_custom
   ```
   Registration happens as an import side-effect — if it isn't imported here, the flavor
   is invisible to training.

6. **Reference it from a config TOML**: `[model] name = "<name>_custom"` and
   `flavor = "7B"` (or whichever flavor key you defined).

### 2.3 Things to watch (lessons from OLMo3 / Qwen3)

- **Custom-op registration order.** If the model registers a custom op (e.g.
  `ring_flash_attn`), `model.py` **must be imported before** `infra/parallelize.py` —
  see the import ordering comment in `olmo3_custom/__init__.py`. Getting this wrong gives
  "operator not registered" errors at parallelize time.
- **Config defaults vs. overrides.** Don't let `None` job-config values clobber a flavor's
  built-in defaults (this was an explicit fix on this branch). Guard the merge so unset
  TOML keys fall back to the flavor dataclass.
- **Attention type drives collation.** Varlen / ring-attn paths are selected from
  `attn_type`, not a separate flag — keep new models consistent with that so the SFT
  collator picks the right path.
- **CP correctness / buffer-name contract.** If you support context parallelism, validate
  ring-varlen forward **and** backward, and keep it `torch.compile`-safe (zigzag varlen
  needed explicit work here). Also: torchtitan's CP slices the RoPE buffer **only if it is
  named exactly `freqs_cis`** (`train.py:488`) — any other name (e.g. `rope_cache`) is left
  un-sliced and every CP rank applies RoPE for `0..local_seqlen` → wrong Q/K → NaN loss.
  See `docs/cp_varlen_context.md` and `docs/ring_attention_cp_gradient_notes.md`.
- **CP loss normalization.** Under CP, a rank whose label slice is all `-100` yields
  `0/0 = NaN` with `reduction="mean"`. Normalize as `sum(local) / sum_across_cp(global_valid)`
  on every CP path (ring-varlen, SDPA+CP, Liger). This bites both new models and new loss paths.
- **MFU / memory at long seq len.** If the model uses sliding-window attention, make MFU
  accounting reflect it (OLMo3 does sliding-window-aware MFU). For activation checkpointing,
  don't blindly save every `aten.mm` — at long seq len (e.g. 32k) saving each matmul output
  can cost *more* memory than full AC; prefer saving only attention kernels, or module-level
  FFN AC (`selective_ac_option == "ffn"`).
- **Flavor sizing.** Always ship a tiny `debugmodel` flavor for smoke tests in addition
  to the real sizes.
- **Smoke-test it.** Run a short job (e.g. via `scripts/run_reasoning_smoke.py` or the
  SFT entrypoint with `debugmodel`) before launching at scale.

### 2.4 Minimal verification checklist

- [x] `from titan_oellm.models import <name>_custom` succeeds and registers the spec
- [x] `debugmodel` flavor trains a few steps without NaNs
- [x] HF checkpoint loads and saves via the state-dict adapter
- [x] Each intended parallelism axis (TP/FSDP/CP/PP) runs
- [x] If CP: ring-varlen forward+backward numerically match a non-CP reference
- [x] SFT path works with at least one packing strategy

### 2.5 Case study: enabling Qwen3 (commit `9f38c7e`)

Qwen3 already existed as a pretraining model; making it work with the SFT + CP pipeline
took exactly the four glue fixes below — three of which map directly onto the gotchas in
§2.3. No new infrastructure, just per-model adaptation.

| File | Fix | Maps to gotcha |
|---|---|---|
| `model/args.py` | Wire attention dispatch (`attn_mask_type`, `attn_type` / `use_flex_attn`) from job config. Qwen3 was hardcoded to the `"sdpa"` default and asserted when SFT passed `VarlenMetadata`. | §2.3 "attention type drives collation" |
| `model/model.py` | Rename the RoPE buffer `rope_cache` → `freqs_cis`. torchtitan's CP machinery (`train.py:488`) slices the buffer **named exactly `freqs_cis`** per CP rank along the seq dim. Under the old name it wasn't sliced, so every CP rank applied RoPE for positions `0..local_seqlen` → wrong Q/K → NaN loss. | §2.3 "CP correctness / buffer-name contract" |
| `infra/parallelize.py` | Drop `aten.mm.default` and `inductor_compiled_code` from the selective-op AC save list (saving every matmul at long seq len cost more memory than full AC), and add module-level FFN activation checkpointing when `selective_ac_option == "ffn"`. | §2.3 "MFU / memory at long seq len" |
| `scripts/sft.py` | Add an SDPA+CP forward/backward path with **global-valid loss normalization**: `sum(local) / sum_across_cp(global_valid)`. Any CP rank whose label slice is all `-100` otherwise yields `0/0 = NaN` under `reduction="mean"`. Also seed python `random` + `numpy` under deterministic mode. | §2.3 "CP loss normalization" |

**Takeaway:** the buffer-name contract (`freqs_cis`) and the CP all-`-100` NaN are the two
non-obvious traps. Any new model that supports CP will hit both — budget for them up front.

---
